### PR TITLE
Support custom route methods && uniswap v3 multi hop

### PR DESCRIPTION
**Pull request by liu-zhipeng**
- [x] support custom router abi (For Pangolin and Trader Joe swap on avalanche)
- [x] support custom router method names
- [x] support uniswap v3 multihop
- [x] fix some issues
- [x] Unit tests
- [x] readme

### CODE REVIEW

**Comment by liu-zhipeng** [comment]
Hi. I am so excited to contribute to such a great project. I found some small problems while using uniswap sdk. For example, some uniswap forked dexs like pangolin swap changed router functions. especially ETH to AVAX. So i added some changes for developers to use custom router abi and router methods. Also I added some code to support uniswap v3 multihop. I would like you to review my changes if you have time. Thank you again for a great sdk!

**Comment by joshstevens19** [comment]
Hey @liu-zhipeng firstly thanks so much for this!! Help from others is needed and this looks incredible already. 

Only scanned on phone but 1 thing which popped out any new things or featured you added to this please add to the readme doc, for example you done some liquidity work and now supporting multihop adding to the docs is the last part missing I think! 

I will review this tomorrow be great to review with the docs update as well thanks a lot hope this will be your first of many PRs 💪

**Comment by liu-zhipeng** [comment]
> Hey @liu-zhipeng firstly thanks so much for this!! Help from others is needed and this looks incredible already.
> 
> Only scanned on phone but 1 thing which popped out any new things or featured you added to this please add to the readme doc, for example you done some liquidity work and now supporting multihop adding to the docs is the last part missing I think!
> 
> I will review this tomorrow be great to review with the docs update as well thanks a lot hope this will be your first of many PRs 💪

**Review by savasadar — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -464,7 +474,7 @@ export interface TradeContext {
     routeText: string;
     routePathArray: string[];
     uniswapVersion: UniswapVersion;
-    liquidityProviderFee: number;
+    liquidityProviderFee: number[];
```
Comment (joshstevens19) [inline README.md:477]: doesn't seem correct here as in v2 it would still be string

File: README.md [hunk 2]
```diff
@@ -891,8 +901,8 @@ console.log(trade);
   minAmountConvertQuote: '0.014400465273974444',
   maximumSent: null,
   expectedConvertQuote: '0.014730394044348867',
-  liquidityProviderFee: '0.030000000000000000',
-  liquidityProviderFeePercent: 0.003,
+  liquidityProviderFee: ['0.030000000000000000','0.050000000000000000','0.030000000000000000'],
+  liquidityProviderFeePercent: [0.003, 0.005, 0.003],
```
Comment (joshstevens19) [inline README.md:905]: from my memory the liquidity provider fee percent on this level is just 1 percent and if you got many routes you can go then it many objects in the array

File: README.md [hunk 3]
```diff
@@ -891,8 +901,8 @@ console.log(trade);
   minAmountConvertQuote: '0.014400465273974444',
   maximumSent: null,
   expectedConvertQuote: '0.014730394044348867',
-  liquidityProviderFee: '0.030000000000000000',
-  liquidityProviderFeePercent: 0.003,
+  liquidityProviderFee: ['0.030000000000000000','0.050000000000000000','0.030000000000000000'],
```
Comment (joshstevens19) [inline README.md:904]: same as below

File: README.md [hunk 4]
```diff
@@ -959,7 +969,7 @@ console.log(trade);
           '0x1985365e9f78359a9B6AD760e32412f4a445E862',
         ],
         uniswapVersion: 'v2',
-        liquidityProviderFee: 0.003
+        liquidityProviderFee: [0.003, 0.003],
```
Comment (joshstevens19) [inline README.md:972]: breaking change for v2...

File: README.md [hunk 5]
```diff
@@ -1001,7 +1011,7 @@ console.log(trade);
           '0x1985365e9f78359a9B6AD760e32412f4a445E862',
         ],
         uniswapVersion: 'v3',
-        liquidityProviderFee: 0.0005
+        liquidityProviderFee: [0.003, 0.005, 0.003],
```
Comment (joshstevens19) [inline README.md:1014]: if your getting a quote as it is now it picks them best price but also factor in the provider fee i would probably expect 3 rows in this case no?

File: src/common/utils/is-same-address.spec.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
+import { ETH } from '../tokens';
+import { isSameAddress } from './is-same-address';
+
+describe('isSameAddress', () => {
+  it('should return true if its a valid ethereum address type', () => {
```
Comment (joshstevens19) [inline src/common/utils/is-same-address.spec.ts:5]: change descriptions of test

File: src/common/utils/is-same-address.spec.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    expect(isSameAddress('0x45Cd08334aeedd8a06265B2Ae302E3597d8fAA28', '0x45cd08334aeedd8a06265b2ae302e3597d8faa28')).toEqual(
+      true
+    );
+  });
+
+  it('should return false when one of them is a invalid address', () => {
```
Comment (joshstevens19) [inline src/common/utils/is-same-address.spec.ts:11]: same

File: src/common/utils/is-same-address.spec.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    expect(isSameAddress('0xC1912fEE45d61C87Cc5EA59DaE31190FFFFf232d', '0xaaaaa')).toEqual(
+      false
+    );
+  });
+
+  it('should return false when both of them are invalid address', () => {
+    expect(isSameAddress('0x0', '0x1')).toEqual(false);
```
Comment (joshstevens19) [inline src/common/utils/is-same-address.spec.ts:18]: same

File: src/factories/pair/models/current-trade-context.ts [hunk 9]
```diff
@@ -8,7 +8,7 @@ export interface CurrentTradeContext {
   quoteDirection: TradeDirection;
   fromToken: Token;
   toToken: Token;
-  liquidityProviderFee: string;
+  liquidityProviderFee: string[];
```
Comment (joshstevens19) [inline src/factories/pair/models/current-trade-context.ts:11]: need to have a bigger discussion here should be 1 object per liquidity provider fee i think... and the top one have the best one! the idea is the quote gives you everything you need to execute the trade

File: src/factories/pair/models/trade-context.ts [hunk 10]
```diff
@@ -12,8 +12,8 @@ export interface TradeContext {
   minAmountConvertQuote: string | null;
   maximumSent: string | null;
   expectedConvertQuote: string;
-  liquidityProviderFee: string;
-  liquidityProviderFeePercent: number;
+  liquidityProviderFee: string[];
+  liquidityProviderFeePercent: number[];
```
Comment (joshstevens19) [inline src/factories/pair/models/trade-context.ts:16]: again arrays here v2 wont be an array

File: src/factories/pair/uniswap-pair.factory.ts [hunk 11]
```diff
@@ -333,14 +333,13 @@ export class UniswapPairFactory {
           ? null
           : bestRouteQuote.expectedConvertQuoteOrTokenAmountInMaxWithSlippage,
       expectedConvertQuote: bestRouteQuote.expectedConvertQuote,
-      liquidityProviderFee:
+      liquidityProviderFee: bestRouteQuote.liquidityProviderFee.map((f) =>
```
Comment (joshstevens19) [inline src/factories/pair/uniswap-pair.factory.ts:336]: breaking changes for v2.. if many liquidity provider fees the expected convert quote will be different depending on that no? would not expect a loop here i dont think...

File: src/factories/pair/uniswap-pair.factory.ts [hunk 12]
```diff
@@ -404,14 +403,13 @@ export class UniswapPairFactory {
...
```
Comment (joshstevens19) [inline src/factories/pair/uniswap-pair.factory.ts:406]: same as above

File: src/factories/pair/uniswap-pair.factory.ts [hunk 13]
```diff
@@ -470,14 +468,13 @@ export class UniswapPairFactory {
...
```
Comment (joshstevens19) [inline src/factories/pair/uniswap-pair.factory.ts:471]: same as above

File: src/factories/router/models/token-routes.ts [hunk 14]
```diff
@@ -1,9 +1,9 @@
-import { Token } from '../../token/models/token';
+import { Pool, Token } from "../../token/models/token";
 
 export interface TokenRoutes {
   token: Token;
   pairs: {
-    fromTokenPairs?: Token[] | undefined;
-    toTokenPairs?: Token[] | undefined;
+    fromTokenPairs?: Pool[] | undefined;
```
Comment (joshstevens19) [inline src/factories/router/models/token-routes.ts:6]: what about v2 support?

**Comment by mburger81** [comment]
@joshstevens19 @liu-zhipeng 
when do you think you can find a solution for the changes requested and release this?

**Comment by liu-zhipeng** [comment]
> @joshstevens19 @liu-zhipeng when do you think you can find a solution for the changes requested and release this?

Once @joshstevens19  confirms this, we can release this, cus all features are working properly now.

**Comment by liu-zhipeng** [comment]
That is no problem at all to return `liquidityProviderFee` as array. If there is some issues on the apps which are using old version. they MUST update to support uniswap v3. Because each uniswap V3 pairs have 3 different fee tiers.
I am waiting for @joshstevens19 's reivew yet.

**Comment by liu-zhipeng** [comment]
This PR should be closed? @joshstevens19

**Comment by joshstevens19** [comment]
No PR is good sorry been away I’m going to get to this soon!

**Comment by mburger81** [comment]
@liu-zhipeng why did you close this unmerged?

File: README.md [hunk 15]
```diff
@@ -420,12 +426,16 @@ export interface TradeContext {
...
+  // v2 always = (0.3%) ex [0.03, 0.03]
   // v3 depends on the fee amount sent on that pool
   // - low = 0.05%
   // - medium = 0.3%
   // - high = 1%
-  liquidityProviderFee: string;
+  // For multihop swap
+  // Sequence of Fees 
+  // ex: USDC>USDT>WETH>WBTC => [(USDC-USDT fee), (USDT-WETH fee), (WETH-WBTC fee)]
+  liquidityProviderFee: string[];
```
[the remark this answers is not in the document]
Comment (helios8090) [reply README.md:438]: in v2 `liquidityProviderFee` should be array in this version

[the remark this answers is not in the document]
Comment (helios8090) [reply README.md:438]: @liu-zhipeng am i correct?

**Comment by helios8090** [comment]
when will you release this? @liu-zhipeng

**Comment by mburger81** [comment]
@helios8090 the problem is, @liu-zhipeng can not do the merge and the release
@joshstevens19 sorry dude but its really annoying :(

**Comment by liu-zhipeng** [comment]
> @helios8090 the problem is, @liu-zhipeng can not do the merge and the release @joshstevens19 sorry dude but its really annoying :(

I have no permission to merge.

**Comment by mburger81** [comment]
@liu-zhipeng I know, for that reason I asked you to fork, merge and release your own repository, just temporally until @joshstevens19 will merge it

we are trying to make webcomponents for the community using simple-uniswap to deliver a well working, free, fast, working everywhere solution
but we are stuck on waiting this merge since e 5 weeks

**Comment by joshstevens19** [comment]
Guys super sorry I’ve been away and it’s super hard for me to see these sometimes. The main problem with the PR is it’s a breaking interface change to the liquidity interface which means if we release it everyones integration will break who use to think that was a string. I’m going to set a note for tomorrow AM before I do anything else to do another PR review on this. I did do a PR review but the things I asked still is missing. We can’t do breaking changes in the interface we can ofc add a new field and depreciate the old one for v1 and v2 use only but if you could make it so v2 interface is NONE breaking then I’m happy to get this in asap @liu-zhipeng

**Comment by joshstevens19** [comment]
Remind me set.. tomorrow AM 9:45 BST! will do another review of it and highlight the main issues. Sorry for the delay!

**Comment by mburger81** [comment]
Yeah that was what I'm saying, there is this breaking change you already told us
for that reason I was wondering why you wrote later, anything is fine, because it isn't

I Just ask you to maybe integrate this changes also on the swap ui you provide

THX a lot!

**Comment by joshstevens19** [comment]
Well it is - you just expose the field in a new property in the interface and write docs around it why, it’s a very easy fix in this PR!

File: src/factories/pair/models/current-trade-context.ts (hunk 9 above)
Comment (joshstevens19) [reply src/factories/pair/models/current-trade-context.ts:11]: This here! This needs to be a new field if it has to be an array so it doesn’t do a breaking change to the interface

File: src/factories/pair/models/trade-context.ts (hunk 10 above)
Comment (joshstevens19) [reply src/factories/pair/models/trade-context.ts:16]: Same as above breaking change in the trade context ..

**Review by joshstevens19 — CHANGES_REQUESTED** [review]

**Comment by mburger81** [comment]
@liu-zhipeng nice!

**Comment by joshstevens19** [comment]
this looks better @liu-zhipeng can you just update the readme docs to show the new field and explain it so people understand?

**Comment by liu-zhipeng** [comment]
> I have already updated readme.

**Comment by liu-zhipeng** [comment]
https://github.com/joshstevens19/simple-uniswap-sdk/pull/45/commits/7a178a7e9618ef32258b2e6daaa8d076189f183d

**Review by joshstevens19 — APPROVED** [review]
need to run 1 manual test later if good will merge and deploy.. thanks for doing this and sorry it ended up taking longer then we wanted. @liu-zhipeng your a legend!

**Comment by liu-zhipeng** [comment]
Thank you for your approval. @joshstevens19  And I am super happy if my work can be helpful for others.

**Comment by joshstevens19** [comment]
@liu-zhipeng so done manually tests everything about done just one thing I spotted is when on a v2 trade the `liquidityProviderFeesV3` is populated it should be an empty array [] could you do that then we can merge and deploy!


image

**Comment by liu-zhipeng** [comment]
> @liu-zhipeng so done manually tests everything works well, just one thing I spotted is when on a v2 trade the `liquidityProviderFeesV3` is populated it should be an empty array [] could you do that then we can merge and deploy!
> 
> image

I thought that was more consistent. anyway updated finally!

**Comment by joshstevens19** [comment]
amazing! Will deploy tomorrow @liu-zhipeng thanks again

**Comment by liu-zhipeng** [comment]
Awesome!

**Comment by joshstevens19** [comment]
https://github.com/joshstevens19/simple-uniswap-sdk/releases/tag/3.7.0 - its now deployed

**Pull request merged**